qiskit/transpiler/passes/basis/unroll_3q_or_more.py

```diff
@@ -70,11 +71,23 @@ def run(self, dag):
             elif self.basis_gates is not None and node.name in self.basis_gates:
                 continue
 
+            params = node.op.params
+            name = node.op.name
+            if not params:
+                cached_decomp = self._cache.get(name, None)
```

Comment:
What about gates that have non-unique names, such as `Isometry` or `PauliEvolutionGate`? 🤔

Reply:
Sigh, I always forget about those gates. But I think this should still be sound because they all have parameters set so this won't cache them. But if you can think of an example this isn't sound then I have another idea to get this result (move to the basis translator use that instead).

Reply:
I think the biggest risk is with things like `QFT` that aren't parametrised, but are variadic over the qubit count.

Reply:
Well I guess there are two paths here then, either we can add a qargs `len()` or a tuple of qargs to the cache key and assume with a given name and qubits it's the same gate. Or alternatively we can abandon this caching approach and move to the doing unroll3q to us the basis translator instead (by adding a equiv library option to this pass and a min qubits arg to the basis translator). I'm thinking the basis translator approach will be the robust one, because the use case I'm concerned with here is if someone submits a circuit with 200k toffolis we spend a huge amount of time unrolling the same toffoli over and over again.

Reply:
Yeah, I feel like this should be a bit more of a unified part of the basis translator.  I actually think we might want to consider what a transpiler pipeline that has basis translation before routing would look like, so that we don't need the 3+q split at all; I could imagine a `Target`-aware router that synthesises the swaps into something allowable immediately on output, such that basis-translation afterwards isn't necessary.